REGRESSION (292627@main): Doubled text on maps.google.com.

mattwoodrow opened the pull request:
#### 0d37fd4f8a3e6af20ad6a1e11a79f191ab8ccaa7
<pre>
REGRESSION (292627@main): Doubled text on maps.google.com.
<a href="https://bugs.webkit.org/show_bug.cgi?id=291916">https://bugs.webkit.org/show_bug.cgi?id=291916</a>
&lt;<a href="https://rdar.apple.com/149594692">rdar://149594692</a>&gt;

Reviewed by Simon Fraser.

Layers that aren&apos;t self painting don&apos;t hold cached repaint rects, but they
should still hold a cached repaint container.

RenderLayer::computeRepaintRects handled this correctly, but
recursiveUpdateLayerPositions duplicates the logic but without setting the
repaint container.

Remove the duplicated path (and fix assertions appropriately) so that we instead
just call into computeRepaintRects (like we do for self-painting layers) and
have the code there handle it correctly.

Fixes the bug where a previously not-self-painting becomes self painting, and
the repaint got sent to the null cached repaint container (the view).

Testcase has a layer that becomes self painting (due to composited overflow
scrolling), and has a repaint container that isn&apos;t the view. Check the repaint
issued doesn&apos;t end up at the view.

Also extends the layer positions assertions code to dump the tree (in debug
builds), with the failing node marked, similar to how
RenderTreeNeedsLayoutChecker works.

* LayoutTests/compositing/repaint/change-notselfpainting-to-scroller-expected.txt: Added.
* LayoutTests/compositing/repaint/change-notselfpainting-to-scroller.html: Added.
* LayoutTests/svg/compositing/transform-change-repainting-viewBox-repaintRects-expected.txt:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::outputLayerPositionTreeRecursive):
(WebCore::showLayerPositionTree):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::layerWillBeRemoved):

Canonical link: <a href="https://commits.webkit.org/294225@main">https://commits.webkit.org/294225@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ffb050f82d084e5ba16690646b77a36c39ab43a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101200 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20862 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11165 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/106348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51827 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103241 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21171 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29356 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/106348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34111 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104207 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16322 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91412 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/106348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/16139 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/9423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/51175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/86027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9496 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108704 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28328 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86052 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28690 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87615 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85600 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21779 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30323 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/8042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/22427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28258 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33529 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28070 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31390 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29628 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -1532,6 +1532,6 @@ void showLayerTree(const WebCore::RenderLayer*);
 void showPaintOrderTree(const WebCore::RenderLayer*);
 void showPaintOrderTree(const WebCore::RenderObject*);
-void showLayerPositionTree(const WebCore::RenderLayer*);
+void showLayerPositionTree(const WebCore::RenderLayer*, const WebCore::RenderLayer* = nullptr);
```
The parameters need names, now that they are no longer obvious.

mattwoodrow committed:
REGRESSION (292627@main): Doubled text on maps.google.com.
https://bugs.webkit.org/show_bug.cgi?id=291916
<rdar://149594692>

Reviewed by Simon Fraser.

Layers that aren't self painting don't hold cached repaint rects, but they
should still hold a cached repaint container.

RenderLayer::computeRepaintRects handled this correctly, but
recursiveUpdateLayerPositions duplicates the logic but without setting the
repaint container.

Remove the duplicated path (and fix assertions appropriately) so that we instead
just call into computeRepaintRects (like we do for self-painting layers) and
have the code there handle it correctly.

Fixes the bug where a previously not-self-painting becomes self painting, and
the repaint got sent to the null cached repaint container (the view).

Testcase has a layer that becomes self painting (due to composited overflow
scrolling), and has a repaint container that isn't the view. Check the repaint
issued doesn't end up at the view.

Also extends the layer positions assertions code to dump the tree (in debug
builds), with the failing node marked, similar to how
RenderTreeNeedsLayoutChecker works.

* LayoutTests/compositing/repaint/change-notselfpainting-to-scroller-expected.txt: Added.
* LayoutTests/compositing/repaint/change-notselfpainting-to-scroller.html: Added.
* LayoutTests/svg/compositing/transform-change-repainting-viewBox-repaintRects-expected.txt:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::outputLayerPositionTreeRecursive):
(WebCore::showLayerPositionTree):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::layerWillBeRemoved):

Canonical link: https://commits.webkit.org/294225@main
